core/database/foxx/api/repo_router_new.js

```diff
@@ -0,0 +1,283 @@
+            ...(!req.queryParams.all && client && { admin: client._id }),
+        };
+
+        const listResult = RepositoryOps.list(filter);
+        if (!listResult.ok) throw [listResult.error.code, listResult.error.message];
+
+        res.send(listResult.value.map(cleanRepoData));
+    })
+    .queryParam("client", joi.string().allow("").optional(), "Client ID")
+    .queryParam("details", joi.boolean().optional(), "Show additional record details")
+    .queryParam("all", joi.boolean().optional(), "List all repos (requires admin)")
+    .queryParam(
+        "type",
+        joi.string().valid("globus", "metadata_only").optional(),
+        "Filter by repository type",
+    )
+    .summary("List repositories")
+    .description("List repositories administered by the client.");
+
+// View repository
+router
+    .get("/view", (req, res) => {
+        try {
+            const findResult = RepositoryOps.find(req.queryParams.id);
+            if (!findResult.ok) throw [findResult.error.code, findResult.error.message];
+
+            const repo = cleanRepoData(findResult.value);
+            repo.admins = g_db.admin
+                .byExample({ _from: repo.id })
+                .toArray()
+                .map((admin) => admin._to);
+
+            res.send([repo]);
+        } catch (e) {
+            g_lib.handleException(e, res);
+        }
+    })
+    .queryParam("id", joi.string().required(), "Repository ID")
+    .summary("View repository")
+    .description("View repository details.");
+
+// Create repository
+router
+    .post("/create", (req, res) => {
+        try {
+            g_db._executeTransaction({
+                collections: { read: ["u"], write: ["repo", "admin"] },
+                action: () => {
+                    const client = g_lib.getUserFromClientID(req.queryParams.client);
+                    if (!client.is_admin) throw g_lib.ERR_PERM_DENIED;
+
+                    const config = {
+                        ...req.body,
+                        ...(req.body.type === RepositoryType.GLOBUS && {
+                            pub_key: req.body.pub_key,
+                            address: req.body.address,
+                            endpoint: req.body.endpoint,
+                            path: req.body.path,
+                            exp_path: req.body.exp_path,
+                            domain: req.body.domain,
+                        }),
+                    };
+
+                    const createResult = createRepositoryByType(config);
+                    if (!createResult.ok)
+                        throw [createResult.error.code, createResult.error.message];
+
+                    const repository = createResult.value;
+
+                    req.body.admins.forEach((admin) => {
+                        if (!g_db._exists(admin))
+                            throw [g_lib.ERR_NOT_FOUND, `User ${admin} not found`];
+                    });
```

Comment:
It might make sense to validate incoming request arguments before doing much else, i.e. before calling createRepositoryByType or creating config arguments. Could be a nice pattern to follow, place all request validation near the top of the route.